compiler/incremental-compilation-impl/src/org/jetbrains/kotlin/incremental/IncrementalCompilerRunner.kt

```diff
@@ -75,192 +86,210 @@ abstract class IncrementalCompilerRunner<
-            }
+    private fun getClasspathAbiSnapshot(args: Args): Map<String, AbiSnapshot> {
+        return reporter.measure(BuildTime.SET_UP_ABI_SNAPSHOTS) {
+            setupJarDependencies(args, reporter)
         }
-        return rebuild(rebuildReason, allSourceFiles, args, messageCollector, providedChangedFiles, projectDir, classpathAbiSnapshot)
     }
 
     /**
-     * Deletes output files and contents of output directories on rebuild, including `@LocalState` files/directories.
+     * Deletes the contents of the given directories (not the directories themselves).
      *
      * If the directories do not yet exist, they will be created.
      */
-    private fun cleanOutputsAndLocalStateOnRebuild(args: Args) {
-        // Use Set as additionalOutputFiles may already contain destinationDir and workingDir
-        val outputFiles = setOf(destinationDir(args), workingDir) + additionalOutputFiles
-
-        reporter.debug { "Cleaning outputs on rebuild" }
-        outputFiles.forEach {
+    private fun cleanOrCreateDirectories(outputDirs: Collection<File>) {
+        outputDirs.toSet().forEach {
             when {
-                it.isDirectory -> {
-                    reporter.debug { "  Deleting contents of directory '${it.path}'" }
-                    it.cleanDirectoryContents()
-                }
-                it.isFile -> {
-                    reporter.debug { "  Deleting file '${it.path}'" }
-                    it.deleteRecursivelyOrThrow()
-                }
+                it.isDirectory -> it.deleteDirectoryContents()
+                it.isFile -> "Expected a directory but found a regular file: ${it.path}"
+                else -> it.createDirectory()
             }
         }
     }
 
-    private fun sourcesToCompile(
-        caches: CacheManager,
-        changedFiles: ChangedFiles,
-        args: Args,
-        messageCollector: MessageCollector,
-        dependenciesAbiSnapshots: Map<String, AbiSnapshot>
-    ): CompilationMode =
-        when (changedFiles) {
-            is ChangedFiles.Known -> calculateSourcesToCompile(caches, changedFiles, args, messageCollector, dependenciesAbiSnapshots)
-            is ChangedFiles.Unknown -> CompilationMode.Rebuild(BuildAttribute.UNKNOWN_CHANGES_IN_GRADLE_INPUTS)
-            is ChangedFiles.Dependencies -> error("Unexpected ChangedFiles type (ChangedFiles.Dependencies)")
-        }
-
-    private fun calculateSourcesToCompile(
-        caches: CacheManager, changedFiles: ChangedFiles.Known, args: Args, messageCollector: MessageCollector,
-        abiSnapshots: Map<String, AbiSnapshot>
-    ): CompilationMode =
-        reporter.measure(BuildTime.IC_CALCULATE_INITIAL_DIRTY_SET) {
-            calculateSourcesToCompileImpl(caches, changedFiles, args, messageCollector, abiSnapshots)
+    private fun getChangedFiles(providedChangedFiles: ChangedFiles?, allSourceFiles: List<File>, caches: CacheManager): ChangedFiles.Known {
```

Comment:
Maybe just inline are make a local function? The fact there is ordering assumption WRT handling ChangedFiles.Unknown makes it a bit tricky to understand what is going on.